File: server/routes/projects.js
const express = require('express');
const { body, param, query, validationResult } = require('express-validator');
const { Op } = require('sequelize');
const { Project, ProjectMember, User, Task } = require('../models');
const auth = require('../middleware/auth');
const roleCheck = require('../middleware/roleCheck');

const router = express.Router();

// All project routes require authentication
router.use(auth);

// ──────────────────────────────────────────────
// POST /api/projects — Create a new project
// ──────────────────────────────────────────────
router.post(
  '/',
  [
    body('name')
      .trim()
      .notEmpty()
      .withMessage('Project name is required')
      .isLength({ max: 255 })
      .withMessage('Project name must be at most 255 characters'),
    body('description')
      .optional()
      .trim(),
  ],
  async (req, res, next) => {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(400).json({ message: errors.array().map((e) => e.msg).join('. ') });
      }

      const { name, description } = req.body;

      const project = await Project.create({
        name,
        description: description || '',
        created_by: req.user.id,
      });

      // Creator automatically becomes an admin member
      await ProjectMember.create({
        project_id: project.id,
        user_id: req.user.id,
        role: 'admin',
      });

      // Reload with associations
      const fullProject = await Project.findByPk(project.id, {
        include: [
          { model: User, as: 'creator', attributes: ['id', 'name', 'email'] },
          {
            model: ProjectMember,
            as: 'projectMembers',
            include: [{ model: User, as: 'user', attributes: ['id', 'name', 'email'] }],
          },
        ],
      });

      res.status(201).json({ project: fullProject });
    } catch (error) {
      next(error);
    }
  }
);

// ──────────────────────────────────────────────
// GET /api/projects — List projects the user belongs to
// ──────────────────────────────────────────────
router.get('/', async (req, res, next) => {
  try {
    const memberships = await ProjectMember.findAll({
      where: { user_id: req.user.id },
      attributes: ['project_id', 'role'],
    });

    const projectIds = memberships.map((m) => m.project_id);

    const projects = await Project.findAll({
      where: { id: projectIds },
      include: [
        { model: User, as: 'creator', attributes: ['id', 'name', 'email'] },
        {
          model: ProjectMember,
          as: 'projectMembers',
          include: [{ model: User, as: 'user', attributes: ['id', 'name', 'email'] }],
        },
      ],
      order: [['created_at', 'DESC']],
    });

    // Attach the current user's role to each project
    const projectsWithRole = projects.map((p) => {
      const membership = memberships.find((m) => m.project_id === p.id);
      return {
        ...p.toJSON(),
        currentUserRole: membership ? membership.role : null,
      };
    });

    res.json({ projects: projectsWithRole });
  } catch (error) {
    next(error);
  }
});

// ──────────────────────────────────────────────
// GET /api/projects/:projectId — Get single project
// ──────────────────────────────────────────────
router.get(
  '/:projectId',
  [param('projectId').isUUID().withMessage('Invalid project ID')],
  roleCheck('admin', 'member'),
  async (req, res, next) => {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(400).json({ message: errors.array().map((e) => e.msg).join('. ') });
      }

      const project = await Project.findByPk(req.params.projectId, {
        include: [
          { model: User, as: 'creator', attributes: ['id', 'name', 'email'] },
          {
            model: ProjectMember,
            as: 'projectMembers',
            include: [{ model: User, as: 'user', attributes: ['id', 'name', 'email'] }],
          },
        ],
      });

      if (!project) {
        return res.status(404).json({ message: 'Project not found' });
      }

      res.json({
        project: {
          ...project.toJSON(),
          currentUserRole: req.membership.role,
        },
      });
    } catch (error) {
      next(error);
    }
  }
);

// ──────────────────────────────────────────────
// PUT /api/projects/:projectId — Update project (admin only)
// ──────────────────────────────────────────────
router.put(
  '/:projectId',
  [
    param('projectId').isUUID().withMessage('Invalid project ID'),
    body('name')
      .optional()
      .trim()
      .notEmpty()
      .withMessage('Project name cannot be empty')
      .isLength({ max: 255 })
      .withMessage('Project name must be at most 255 characters'),
    body('description')
      .optional()
      .trim(),
  ],
  roleCheck('admin'),
  async (req, res, next) => {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(400).json({ message: errors.array().map((e) => e.msg).join('. ') });
      }

      const project = await Project.findByPk(req.params.projectId);
      if (!project) {
        return res.status(404).json({ message: 'Project not found' });
      }

      const { name, description } = req.body;
      if (name !== undefined) project.name = name;
      if (description !== undefined) project.description = description;

      await project.save();

      const updatedProject = await Project.findByPk(project.id, {
        include: [
          { model: User, as: 'creator', attributes: ['id', 'name', 'email'] },
          {
            model: ProjectMember,
            as: 'projectMembers',
            include: [{ model: User, as: 'user', attributes: ['id', 'name', 'email'] }],
          },
        ],
      });

      res.json({ project: updatedProject });
    } catch (error) {
      next(error);
    }
  }
);

// ──────────────────────────────────────────────
// DELETE /api/projects/:projectId — Delete project (admin only)
// ──────────────────────────────────────────────
router.delete(
  '/:projectId',
  [param('projectId').isUUID().withMessage('Invalid project ID')],
  roleCheck('admin'),
  async (req, res, next) => {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(400).json({ message: errors.array().map((e) => e.msg).join('. ') });
      }

      const project = await Project.findByPk(req.params.projectId);
      if (!project) {
        return res.status(404).json({ message: 'Project not found' });
      }

      // Delete related records first
      await Task.destroy({ where: { project_id: project.id } });
      await ProjectMember.destroy({ where: { project_id: project.id } });
      await project.destroy();

      res.json({ message: 'Project deleted successfully' });
    } catch (error) {
      next(error);
    }
  }
);

// ──────────────────────────────────────────────
// GET /api/projects/:projectId/member-suggestions?q=...
// Search users by name/email who are not already members (admin only)
// ──────────────────────────────────────────────
router.get(
  '/:projectId/member-suggestions',
  [
    param('projectId').isUUID().withMessage('Invalid project ID'),
    query('q')
      .trim()
      .isLength({ min: 1, max: 100 })
      .withMessage('Search query is required'),
  ],
  roleCheck('admin'),
  async (req, res, next) => {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(400).json({ message: errors.array().map((e) => e.msg).join('. ') });
      }

      const searchTerm = req.query.q.trim();

      const existingMembers = await ProjectMember.findAll({
        where: { project_id: req.params.projectId },
        attributes: ['user_id'],
      });
      const existingMemberIds = existingMembers.map((m) => m.user_id);

      const users = await User.findAll({
        where: {
          id: { [Op.notIn]: existingMemberIds },
          [Op.or]: [
            { name: { [Op.iLike]: `%${searchTerm}%` } },
            { email: { [Op.iLike]: `%${searchTerm}%` } },
          ],
        },
        attributes: ['id', 'name', 'email'],
        order: [['name', 'ASC']],
        limit: 8,
      });

      res.json({ users });
    } catch (error) {
      next(error);
    }
  }
);

// ──────────────────────────────────────────────
// POST /api/projects/:projectId/members — Add member (admin only)
// ──────────────────────────────────────────────
router.post(
  '/:projectId/members',
  [
    param('projectId').isUUID().withMessage('Invalid project ID'),
    body('email')
      .trim()
      .isEmail()
      .withMessage('A valid email is required')
      .normalizeEmail(),
    body('role')
      .optional()
      .isIn(['admin', 'member'])
      .withMessage('Role must be admin or member'),
  ],
  roleCheck('admin'),
  async (req, res, next) => {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(400).json({ message: errors.array().map((e) => e.msg).join('. ') });
      }

      const { email, role } = req.body;

      const userToAdd = await User.findOne({ where: { email } });
      if (!userToAdd) {
        return res.status(404).json({ message: 'User not found with that email' });
      }

      // Check if already a member
      const existing = await ProjectMember.findOne({
        where: { project_id: req.params.projectId, user_id: userToAdd.id },
      });
      if (existing) {
        return res.status(409).json({ message: 'User is already a member of this project' });
      }

      const membership = await ProjectMember.create({
        project_id: req.params.projectId,
        user_id: userToAdd.id,
        role: role || 'member',
      });

      const fullMembership = await ProjectMember.findByPk(membership.id, {
        include: [{ model: User, as: 'user', attributes: ['id', 'name', 'email'] }],
      });

      res.status(201).json({ member: fullMembership });
    } catch (error) {
      next(error);
    }
  }
);

// ──────────────────────────────────────────────
// DELETE /api/projects/:projectId/members/:userId — Remove member (admin only)
// ──────────────────────────────────────────────
router.delete(
  '/:projectId/members/:userId',
  [
    param('projectId').isUUID().withMessage('Invalid project ID'),
    param('userId').isUUID().withMessage('Invalid user ID'),
  ],
  roleCheck('admin'),
  async (req, res, next) => {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(400).json({ message: errors.array().map((e) => e.msg).join('. ') });
      }

      const { projectId, userId } = req.params;

      // Prevent admin from removing themselves if they are the project creator
      const project = await Project.findByPk(projectId);
      if (project && project.created_by === userId) {
        return res.status(400).json({ message: 'Cannot remove the project creator' });
      }

      const membership = await ProjectMember.findOne({
        where: { project_id: projectId, user_id: userId },
      });

      if (!membership) {
        return res.status(404).json({ message: 'Member not found in this project' });
      }

      // Unassign tasks that were assigned to this user in this project
      await Task.update(
        { assigned_to: null },
        { where: { project_id: projectId, assigned_to: userId } }
      );

      await membership.destroy();

      res.json({ message: 'Member removed successfully' });
    } catch (error) {
      next(error);
    }
  }
);

module.exports = router;

File: server/routers/projects.py
"""Project routes — CRUD, members, member-suggestions."""

from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select, or_
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from server.database import get_db
from server.models import Project, ProjectMember, User, Task
from server.schemas import (
    ProjectCreate,
    ProjectUpdate,
    MemberAdd,
    UserOut,
    ProjectMemberOut,
)
from server.auth import get_current_user, require_project_role

router = APIRouter(prefix="/api/projects", tags=["projects"])


# ── helpers ──────────────────────────────────────

def _serialize_user(u: User) -> dict:
    return {"id": str(u.id), "name": u.name, "email": u.email}


def _serialize_member(pm: ProjectMember) -> dict:
    return {
        "id": str(pm.id),
        "project_id": str(pm.project_id),
        "user_id": str(pm.user_id),
        "role": pm.role,
        "joined_at": pm.joined_at.isoformat() if pm.joined_at else None,
        "user": _serialize_user(pm.user) if pm.user else None,
    }


def _serialize_project(p: Project) -> dict:
    return {
        "id": str(p.id),
        "name": p.name,
        "description": p.description or "",
        "created_by": str(p.created_by),
        "created_at": p.created_at.isoformat() if p.created_at else None,
        "creator": _serialize_user(p.creator) if p.creator else None,
        "projectMembers": [_serialize_member(pm) for pm in (p.project_members or [])],
    }


async def _load_project(db, project_id):
    result = await db.execute(
        select(Project)
        .options(
            selectinload(Project.creator),
            selectinload(Project.project_members).selectinload(ProjectMember.user),
        )
        .where(Project.id == project_id)
    )
    return result.scalar_one_or_none()


# ── POST /api/projects ───────────────────────────
@router.post("", status_code=201)
async def create_project(
    body: ProjectCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    project = Project(
        name=body.name.strip(),
        description=(body.description or "").strip(),
        created_by=current_user.id,
    )
    db.add(project)
    await db.flush()

    # Creator becomes admin
    membership = ProjectMember(
        project_id=project.id,
        user_id=current_user.id,
        role="admin",
    )
    db.add(membership)
    await db.commit()

    full = await _load_project(db, project.id)
    return {"project": _serialize_project(full)}


# ── GET /api/projects ────────────────────────────
@router.get("")
async def list_projects(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    # Get user's memberships
    mem_result = await db.execute(
        select(ProjectMember).where(ProjectMember.user_id == current_user.id)
    )
    memberships = mem_result.scalars().all()
    project_ids = [m.project_id for m in memberships]
    role_map = {m.project_id: m.role for m in memberships}

    if not project_ids:
        return {"projects": []}

    result = await db.execute(
        select(Project)
        .options(
            selectinload(Project.creator),
            selectinload(Project.project_members).selectinload(ProjectMember.user),
        )
        .where(Project.id.in_(project_ids))
        .order_by(Project.created_at.desc())
    )
    projects = result.scalars().all()

    out = []
    for p in projects:
        d = _serialize_project(p)
        d["currentUserRole"] = role_map.get(p.id)
        out.append(d)

    return {"projects": out}


# ── GET /api/projects/{project_id} ───────────────
@router.get("/{project_id}")
async def get_project(
    project_id: UUID,
    membership: ProjectMember = Depends(require_project_role("admin", "member")),
    db: AsyncSession = Depends(get_db),
):
    full = await _load_project(db, project_id)
    if not full:
        raise HTTPException(status_code=404, detail="Project not found")

    d = _serialize_project(full)
    d["currentUserRole"] = membership.role
    return {"project": d}


# ── PUT /api/projects/{project_id} ───────────────
@router.put("/{project_id}")
async def update_project(
    project_id: UUID,
    body: ProjectUpdate,
    membership: ProjectMember = Depends(require_project_role("admin")),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(select(Project).where(Project.id == project_id))
    project = result.scalar_one_or_none()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")

    if body.name is not None:
        project.name = body.name.strip()
    if body.description is not None:
        project.description = body.description.strip()

    await db.commit()

    full = await _load_project(db, project_id)
    return {"project": _serialize_project(full)}


# ── DELETE /api/projects/{project_id} ────────────
@router.delete("/{project_id}")
async def delete_project(
    project_id: UUID,
    membership: ProjectMember = Depends(require_project_role("admin")),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(select(Project).where(Project.id == project_id))
    project = result.scalar_one_or_none()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")

    # Delete tasks, then members, then project
    await db.execute(
        Task.__table__.delete().where(Task.project_id == project_id)
    )
    await db.execute(
        ProjectMember.__table__.delete().where(ProjectMember.project_id == project_id)
    )
    await db.delete(project)
    await db.commit()

    return {"message": "Project deleted successfully"}


# ── GET /api/projects/{project_id}/member-suggestions ──
@router.get("/{project_id}/member-suggestions")
async def member_suggestions(
    project_id: UUID,
    q: str = "",
    membership: ProjectMember = Depends(require_project_role("admin")),
    db: AsyncSession = Depends(get_db),
):
    search_term = q.strip()
    if not search_term or len(search_term) > 100:
        raise HTTPException(status_code=400, detail="Search query is required")

    # Get existing member IDs
    mem_result = await db.execute(
        select(ProjectMember.user_id).where(ProjectMember.project_id == project_id)
    )
    existing_ids = [row[0] for row in mem_result.all()]

    query = select(User).where(
        User.id.notin_(existing_ids) if existing_ids else True,
        or_(
            User.name.ilike(f"%{search_term}%"),
            User.email.ilike(f"%{search_term}%"),
        ),
    ).order_by(User.name.asc()).limit(8)

    result = await db.execute(query)
    users = result.scalars().all()

    return {"users": [_serialize_user(u) for u in users]}


# ── POST /api/projects/{project_id}/members ──────
@router.post("/{project_id}/members", status_code=201)
async def add_member(
    project_id: UUID,
    body: MemberAdd,
    membership: ProjectMember = Depends(require_project_role("admin")),
    db: AsyncSession = Depends(get_db),
):
    # Find user by email
    result = await db.execute(
        select(User).where(User.email == body.email.lower())
    )
    user_to_add = result.scalar_one_or_none()
    if not user_to_add:
        raise HTTPException(status_code=404, detail="User not found with that email")

    # Check if already member
    existing = await db.execute(
        select(ProjectMember).where(
            ProjectMember.project_id == project_id,
            ProjectMember.user_id == user_to_add.id,
        )
    )
    if existing.scalar_one_or_none():
        raise HTTPException(
            status_code=409, detail="User is already a member of this project"
        )

    role = body.role if body.role in ("admin", "member") else "member"
    new_member = ProjectMember(
        project_id=project_id,
        user_id=user_to_add.id,
        role=role,
    )
    db.add(new_member)
    await db.commit()
    await db.refresh(new_member)

    # Reload with user relation
    result = await db.execute(
        select(ProjectMember)
        .options(selectinload(ProjectMember.user))
        .where(ProjectMember.id == new_member.id)
    )
    full_member = result.scalar_one()

    return {"member": _serialize_member(full_member)}


# ── DELETE /api/projects/{project_id}/members/{user_id} ──
@router.delete("/{project_id}/members/{user_id}")
async def remove_member(
    project_id: UUID,
    user_id: UUID,
    membership: ProjectMember = Depends(require_project_role("admin")),
    db: AsyncSession = Depends(get_db),
):
    # Prevent removing the project creator
    result = await db.execute(select(Project).where(Project.id == project_id))
    project = result.scalar_one_or_none()
    if project and project.created_by == user_id:
        raise HTTPException(
            status_code=400, detail="Cannot remove the project creator"
        )

    result = await db.execute(
        select(ProjectMember).where(
            ProjectMember.project_id == project_id,
            ProjectMember.user_id == user_id,
        )
    )
    mem = result.scalar_one_or_none()
    if not mem:
        raise HTTPException(
            status_code=404, detail="Member not found in this project"
        )

    # Unassign tasks from the removed member
    from sqlalchemy import update

    await db.execute(
        update(Task)
        .where(Task.project_id == project_id, Task.assigned_to == user_id)
        .values(assigned_to=None)
    )

    await db.delete(mem)
    await db.commit()

    return {"message": "Member removed successfully"}
